README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# elucidate <img src='man/figures/hex-elucidate.png' align="right" height="139" />

## Project Status

<!-- badges: start -->

[![img](https://img.shields.io/badge/Lifecycle-Maturing-007EC6)](https://github.com/bcgov/repomountie/blob/master/doc/lifecycle-badges.md)
[![R-CMD-check](https://github.com/bcgov/elucidate/workflows/R-CMD-check/badge.svg)](https://github.com/bcgov/elucidate/actions)
<!-- badges: end -->

This package is currently maintained by [Craig
Hutton](https://craig.rbind.io/), a Data Scientist working with the
Research Branch of the [British Columbia Ministry of Social Development
& Poverty
Reduction](https://www2.gov.bc.ca/gov/content/governments/organizational-structure/ministries-organizations/ministries/social-development-poverty-reduction).

## Why `elucidate`?

`elucidate` provides a collection of convenience functions to make
exploratory data analysis in R easier and more accessible for
researchers to (among other things):

-   Interrogate data in search of row duplicates with `dupes()` and
    anomalous values with the `counts*` set of functions.

-   Describe data with the `describe*` set of functions for obtaining
    summary statistics, bootstrapping confidence intervals, and
    detecting missing values.

-   Quickly visualise and interact with data representations with the
    `plot_*` set of functions.

Inspired by tidyverse naming conventions, the core functions of
`elucidate` are organized into sets that begin with a common root
(e.g. `describe*`, `plot_*`), since this enables the user to see them
all as suggestions as you are coding in R studio.

Drawing from similar inspiration, many `elucidate` functions are also
designed to accept a data object as the 1st argument and return a data
or plotting object (e.g. ggplot2 or plotly) so they are compatible with
the pipe operator from the
[magrittr](https://magrittr.tidyverse.org/reference/pipe.html) package
for easy integration into data processing pipelines.

For a comprehensive introduction to the package see the vignette via
`vignette("elucidate")`.

## Installation

You can install the development version of elucidate from this
repository with:

``` r
# use the remotes package to install from a github repository

install.packages("remotes") #only run this 1st if you haven't installed remotes before

remotes::install_github("bcgov/elucidate")
```

The authors of `elucidate` acknowledge and express their gratitude to
the authors of the [tidyverse packages](https://www.tidyverse.org/),
[data.table](https://github.com/Rdatatable/data.table), and the
functions of other dependency packages which were used to build
`elucidate`, since without their effort and ingenuity `elucidate` would
mostly have remained a collection of ideas instead of functions.

## Usage

`dupes()` can tell you how many rows are duplicated based on one or more
variables (default is all of them).

``` r
library(elucidate)
#> Loading required package: ggplot2
#> 
#> Attaching package: 'elucidate'
#> The following object is masked from 'package:base':
#> 
#>     mode

#list any number of variables to use when searching for duplicates after the
#data argument
dupes(pdata, d) 
#> Duplicated rows detected! 12000 of 12000 rows in the input data have multiple copies.
#> # A tibble: 12,000 x 11
#>       id d          g     high_low even     y1    y2    x1    x2    x3 n_copies
#>    <int> <date>     <fct> <chr>    <lgl> <dbl> <dbl> <int> <int> <int>    <int>
#>  1     1 2019-01-01 e     high     FALSE  144. 112.     86   199   261     1000
#>  2     2 2019-01-01 c     low      TRUE   262.  91.8    88   104   224     1000
#>  3     3 2019-01-01 d     high     FALSE  224. 120.     29   192   224     1000
#>  4     4 2019-01-01 c     high     TRUE   262. 102.     77   151   254     1000
#>  5     5 2019-01-01 a     low      FALSE  167.  91.6    75   186   258     1000
#>  6     6 2019-01-01 a     low      TRUE   163.  83.6    39   152   293     1000
#>  7     7 2019-01-01 d     low      FALSE  247.  98.9    88   177   202     1000
#>  8     8 2019-01-01 b     high     TRUE   189. 102.     79   163   217     1000
#>  9     9 2019-01-01 e     high     FALSE  156. 110.      4   136   220     1000
#> 10    10 2019-01-01 c     low      TRUE   248.  95.9    63   113   213     1000
#> # ... with 11,990 more rows
#in this case we search for duplicated based on the "d" (date) column        
```

`describe()` a single variable in a data frame or a vector of values.

``` r
#set random generator seed for reproducibility
set.seed(1234)

#using a numeric vector as input
describe(data = rnorm(1:1000, 100, 5))
#> # A tibble: 1 x 14
#>   cases     n    na  p_na  mean    sd    se    p0   p25   p50   p75  p100   skew
#>   <int> <int> <int> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>  <dbl>
#> 1  1000  1000     0     0  99.9  4.99 0.158  83.0  96.6  99.8  103.  116. -0.005
#> # ... with 1 more variable: kurt <dbl>
```

`describe_all()` all variables in a data frame.

``` r
describe_all(pdata)
#> $date
#> # A tibble: 1 x 8
#>   variable cases     n    na  p_na n_unique start      end       
#>   <chr>    <int> <int> <int> <dbl>    <int> <date>     <date>    
#> 1 d        12000 12000     0     0       12 2008-01-01 2019-01-01
#> 
#> $factor
#> # A tibble: 1 x 8
#>   variable cases     n    na  p_na n_unique ordered counts_tb                   
#>   <chr>    <int> <int> <int> <dbl>    <int> <lgl>   <chr>                       
#> 1 g        12000 12000     0     0        5 FALSE   a_2592, b_2460, ..., e_2352~
#> 
#> $character
#> # A tibble: 1 x 9
#>   variable cases     n    na  p_na n_unique min_chars max_chars counts_tb       
#>   <chr>    <int> <int> <int> <dbl>    <int>     <int>     <int> <chr>           
#> 1 high_low 12000 12000     0     0        2         3         4 high_6045, low_~
#> 
#> $logical
#> # A tibble: 1 x 8
#>   variable cases     n    na  p_na n_TRUE n_FALSE p_TRUE
#>   <chr>    <int> <int> <int> <dbl>  <dbl>   <dbl>  <dbl>
#> 1 even     12000 12000     0     0   6000    6000    0.5
#> 
#> $numeric
#> # A tibble: 6 x 15
#>   variable cases     n    na  p_na  mean    sd    se    p0   p25   p50   p75
#>   <chr>    <int> <int> <int> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1 id       12000 12000     0     0 500.  289.  2.64    1   251.   500.  750.
#> 2 y1       12000 12000     0     0 154.   42.7 0.39   69.2 121.   145.  181.
#> 3 y2       12000 12000     0     0 100.   10.1 0.092  60.0  93.4  100.  107.
#> 4 x1       12000 12000     0     0  50.5  28.9 0.264   1    25     50    75 
#> 5 x2       12000 12000     0     0 151.   28.8 0.263 101   126    150   176 
#> 6 x3       12000 12000     0     0 250.   28.9 0.264 201   225    251   276 
#> # ... with 3 more variables: p100 <dbl>, skew <dbl>, kurt <dbl>
```

Use `plot_var()` to produce a class-appropriate
[ggplot2](https://ggplot2.tidyverse.org/) graph of a single variable in
a data frame or a vector of values.

``` r
plot_var(data = rnorm(1:1000, 100, 5)) 
```

<img src="man/figures/README-plot_var-1.png" width="100%" />

``` r
#in this case we get a density plot with a normal density curve added for
#reference (dashed line).
```

To generate class-appropriate `ggplot2` graphs for all variables in a
data frame and combine them into a multiple-panel figure with the
[patchwork](https://patchwork.data-imaginist.com/index.html) package,
use `plot_var_all()`. You can also limit the graphing to a subset of
columns with the “cols” argument, which accepts a character vector of
column names.

``` r
plot_var_all(pdata, cols = c("y1", "y2", "g", "even"))
```

<img src="man/figures/README-plot_var_all-1.png" width="100%" />

``` r
#density plots for numeric variables and bar graphs for categorical variables
```

## Learn more

These examples only highlight a few of the many things `elucidate` can
do. You can learn more from these additional resources:

-   A **[blog
    post](https://craig.rbind.io/post/2020-12-07-asgr-3-0-exploring-data-with-elucidate/)**,
    which was written to help researchers and science trainees learn how
    to use the package for common exploratory data analysis tasks.

-   The **introductory vignette** (`vignette("elucidate")`), which
    provides annotated code examples that demonstrate most of the
    package’s functionality.

## Reporting an Issue

To report bugs/issues or request feature changes, open an
[issue](https://github.com/bcgov/elucidate/issues) for the package
GitHub repo. If raising an issue, *please provide a reproducible
example* ([reprex](https://www.tidyverse.org/help/)) of the problem
you’re encountering.

## Requesting Features and/or Changes

To suggest changes or code improvements, please submit a [pull
request](https://github.com/bcgov/elucidate/pulls).

## License

Copyright 2021 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the “License”); you may
not use this file except in compliance with the License. You may obtain
a copy of the License at:

<http://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an “AS IS” BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


NEWS.md
# elucidate 0.1.1.9001 - May 17th, 2022

* Fixed a bug which led to `describe_all()` calculating and returning uninformative summary statistics for variables in the source data when they are specified as grouping variables. 

# elucidate 0.1.1.9000 - March 2nd, 2022

* Added `describe_na()`. `describe_na_all()`, `plot_na()`, and `plot_na_all()` to facilitate exploration of missing values in data frames. 

* Fixed a bug which caused `describe()` and `describe_ci()` to fail when specifying the "y" argument using a quoted variable name.

* Changed the default output format of `describe_*` functions from "tibble" to "dt" (data.table) for efficiency.

* Changed the default theme of `plot_*` functions to "bw" (`ggplot2::theme_bw()`). 

# elucidate 0.1.0.9005 - January 23rd, 2022

* Fixed a bug which caused `describe()` and `describe_ci()` to fail when attempting to evaluate columns from data.table objects.

# elucidate 0.1.0.9004 - December 13th, 2021

* `describe()` and `describe_all()` now work for "POSIXct" and "POSIXlt" date/time class vectors in addition to the previously supported "Date" class vectors.

# elucidate 0.1.0.9003 - October 31st, 2021

* Fixed a bug that caused `plot_var()` to fail for ordered factors. 

* Fixed a bug that caused `plot_box()` to fail when the optional dot plot layer enhancement was enabled ("dots" = `TRUE`) but no variable was assigned to the x-axis.

* Fixed a bug that prevented `plot_raincloud()` from allowing users to adjust the fill colour of the box plots when the "box_plot" option is enabled.

* Updated the package readme and other documentation.

* Added a package vignette and continuous integration build checks for Windows, MacOS, and Linux.

* Added a "caption" argument to all `plot_*` functions which provides a shortcut for adding a caption to the bottom of the graph that is left-justified by default as per standard scientific communication practice (unlike the `ggplot2` default). All geom-specific `plot_*` functions (not `plot_var*` functions) now also have a "caption_hjust" argument that makes it easy to adjust the horizontal alignment of the caption by adding the appropriate `ggplot2::theme()` layer for you. 

* Changed the name of the `plot_pie()` "title_alignment" argument to "title_hjust" for consistency with other `plot_*` functions, which also now all have a "title_hjust" argument (except for the `plot_var*` functions). The elucidate default title alignment is now 0.5 for centre-justified.

* Rearranged the output of the `*ci()` functions such that the confidence interval lower and upper bound columns are after the summary statistic column.

* The magrittr pipe operator (`%>%`) is no longer imported when `elucidate` is loaded. With the advent of the base R pipe operator (`|>`) in R 4.1 it seemed no longer necessary for `elucidate` to also import a secondary pipe operator, especially since it is already loaded with other common packages like `dplyr`.

# elucidate 0.1.0.9002 - June 28th, 2021

* Fixed a bug that prevented `copies()` and `dupes()` from working correctly when the input data is a data.table. 

* Added citation file.

# elucidate 0.1.0.9001 - June 27th, 2021

* Fixed a bug with `plot_var()` that was preventing the combined box-and-whisker plots + violin plots produced by default for a mix of numeric and categorical variables from generating a plotly graph when the "interactive" argument was set to `TRUE`.

# elucidate 0.1.0.9000 - June 25th, 2021

## major update

### new functions

* Added `plot_var()`, `plot_var_all()`, `plot_var_pairs()`, and `plot_c()` to complete the `plot_*` function set. 

* `plot_var()` is an elucidate alternative to the flexible, but esoteric, `ggplot2::qplot()` function. It can also be thought of as a graphical version of the `describe()` function. `plot_var()` allows you to easily generate a ggplot2 graph for one or two numeric and/or categorical variables using one of four geom-specific `plot_*` functions in a class-appropriate manner, with some default enhancements and a restricted set of arguments:

  - One numeric (classes numeric/integer/date) variable will be graphed with `plot_density()`. By default, a normal density curve will be added as a dashed line. This can be disabled by setting the "dnorm" argument to `FALSE` or the "basic" argument to `TRUE`. 

  - One or two categorical (classes factor/character/logical) variable(s) will be graphed with `plot_bar()`.
  
  - Two numeric variables will be graphed with `plot_scatter()`. By default, a generalized additive model regression line and 95% confidence envelope will also be added to the scatter plot. These extras can be disabled by setting the "regression_line" argument to `FALSE` or the "basic" argument to `TRUE`. 
  
  - A mixture of numeric and categorical variables will be graphed with `plot_box()`. By Default, a `ggplot2::geom_violin()` layer will also be added to enable rapid detection of multi-modal distributions. This additional layer can be disabled by setting the "violin" argument to `FALSE` or the basic argument to `TRUE`.

  - You can also use the "group_var" argument to split any of the above plots by a categorical grouping variable from the same data source, which will be mapped to either fill or colour depending upon the `plot_*` function that is used.
  

R/describe_ci.R
# Copyright 2021 Province of British Columbia
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and limitations under the License.


# describe_ci -------------------------------------------------------------
#' @title
#' Obtain confidence intervals for a summary statistic of a numeric vector.
#'
#' @description describe_ci extends the functionality of \code{\link{stat_ci}}
#'   by allowing you to obtain confidence intervals for a summary statistic
#'   split by any number of grouping variables. Like \code{\link{stat_ci}} you
#'   can specify any function that operates on a numeric variable and returns a
#'   single value (e.g. mean, median, sd, se, etc.), but unlike the other
#'   elucidate package *_ci function this one always returns either a data.table
#'   or tibble (instead of a named vector). Calculations of confidence intervals
#'   for the mean are obtained based on reference to the theoretical
#'   normal/gaussian distribution for speed, otherwise bootstrapping is used,
#'   with options for multicore machines to use parallel processing, which can
#'   speed things up quite a bit for larger samples. \code{\link{stat_ci}} may
#'   be useful instead of describe_ci if you need to pass additional arguments
#'   to the chosen summary statistic function (which is what that function uses
#'   the ... argument for). To get confidence intervals for all numeric
#'   variables in a data frame, use \code{\link{describe_ci_all}} instead.
#'
#' @importFrom data.table data.table
#' @importFrom data.table as.data.table
#' @importFrom data.table dcast
#' @importFrom stats na.omit
#' @importFrom tibble as_tibble
#'
#' @param data Either a numeric vector or a data frame or tibble containing the
#'   numeric vector ("y") to be described and any grouping variables ("...").
#'
#' @param y If the data object is a data.frame, this is the variable for which
#'   you wish to obtain a descriptive summary
#'
#' @param ... If the data object is a data.frame, this special argument accepts
#'   any number of unquoted grouping variable names (also present in the data source)
#'   to use for subsetting, separated by commas (e.g. \code{group_var1,
#'   group_var2})
#'
#' @param stat the unquoted name (e.g. mean, not "mean") of a summary statistic
#'   function to calculate confidence intervals for. Only functions which return
#'   a single value and operate on numeric variables are currently supported.
#'
#' @param replicates The number of bootstrap replicates to use to construct
#'   confidence intervals for statistics other than the sample mean. Default is
#'   2,000, as recommended by Efron & Tibshirani (1993). For publications, or if
#'   you need more precise estimates, more replications (e.g. >= 5,000) are
#'   recommended. N.B. more replications will of course take longer to run. If
#'   you get the error: "estimated adjustment 'a' is NA" when ci_type is set to
#'   "bca" then try again with more replications.
#'
#' @param ci_level The confidence level to use for constructing confidence
#'   intervals. Default is set to \code{ci_level = 0.95} for 95 percent CIs.
#'
#' @param ci_type The type of confidence intervals to calculate from the
#'   bootstrap samples. Most of the options available in the underlying boot.ci
#'   function are implemented (except for studentized intervals): "norm" for an
#'   approximation based on the normal distribution, "perc" for percentile,
#'   "basic" for basic, and "bca" for bias-corrected and accelerated. Percentile
#'   intervals are the default since these are typically sufficient when working
#'   with large data sets (e.g. >= 100,000 rows of data) and are faster to
#'   calculate than BCa intervals. However, BCa intervals (the default for the
#'   more primitive \code{\link{stat_ci}} function) tend to provide the most
#'   accurate/least-biased results (Efron, 1987), particularly for small-medium
#'   sized samples, at the obvious cost of requiring more time to calculate. See
#'   \code{\link[boot]{boot.ci}} for details.
#'
#' @param parallel set to TRUE if you want to use multiple cores or FALSE if you
#'   don't (the default). Note that there is some processing overhead involved
#'   when operating in parallel so speed gains may not be very noticeable for
#'   smaller samples (and may even take longer than sequential processing). Due
#'   to the nature of the underlying parallelization architecture, performance
#'   gains will likely be greater on non-Windows machines that can use the
#'   "multicore" implementation instead of "snow". For obvious reasons this
#'   option only works on machines with more than 1 logical processing core.
#'
#' @param cores If parallel is set to TRUE, this determines the number of cores
#'   to use. To see how many cores are available on your machine, use
#'   parallel::detectCores(). If cores is unspecified the number of available
#'   cores - 1 will be used by default.
#'
#' @param na.rm should missing values be removed before attempting to calculate
#'   the chosen statistic and confidence intervals? Default is TRUE.
#'
#' @param output "tibble" for tibble or "dt" for data.table. Tibble is used as
#'   the default output to facilitate subsequent use/modification of the output
#'   with the tidyverse collection of packages.
#'
#' @author Craig P. Hutton, \email{craig.hutton@@gov.bc.ca}
#'
#' @examples
#'
#' describe_ci(pdata, y1, stat = mean) #the default
#' \dontrun{
#' #using a single core (sequential processing)
#' describe_ci(pdata[1:1000, ], y1, stat = median) #bootstrapped CIs for the median
#'
#' describe_ci(pdata, y1, high_low, stat = mean) #split by a grouping variable
#'
#'
#' #using multiple cores (parallel processing)
#' describe_ci(pdata[1:1000, ], y1, stat = sd, parallel = TRUE, cores = 2)
#' }
#'
#' @references
#' Efron, B. (1987). Better bootstrap confidence intervals. Journal of the
#' American statistical Association, 82(397), 171-185.
#'
#' Efron, B., & Tibshirani, R. J. (1993). An introduction to the bootstrap. New
#' York: Chapman & Hall.
#'
#' @seealso \code{\link[boot]{boot}}, \code{\link[boot]{boot.ci}},
#'   \code{\link{mean_ci}}, \code{\link{median_ci}}, \code{\link{stat_ci}},
#'   \code{\link{describe_ci_all}}
#'
#' @export
describe_ci <- function(data, y = NULL, ..., stat = mean, replicates = 2000,
                        ci_level = 0.95, ci_type = c("perc", "bca", "basic", "norm"), parallel = FALSE, cores = NULL,
                        na.rm = TRUE, output = c("tibble", "dt")) {
  ci_type <- match.arg(ci_type)
  output <- match.arg(output)
  st <- deparse(substitute(stat))
  dt <- data.table::as.data.table(data)

  if(is.error(class(data[[y]]))) {
    y_str <- deparse(substitute(y))
  } else if(!is.character(y) || length(y) > 1){
    stop('If specified, `y` must be a single symbol or character string',
         '\n representing a column in the input data frame supplied to the `data` argument.')
  }

  if(is.numeric(data)){
    if(st == "mean") {
      description <- mean_ci(dt[[1]], replicates = replicates, ci_type = ci_type,
                             ci_level = ci_level, parallel = parallel, cores = cores, na.rm = na.rm)
      description <- data.table::data.table(description[1], description[2], description[3])
      names(description) <- c(st, "lower", "upper")
    } else {
      description <- stat_ci(dt[[1]], stat = stat, replicates = replicates, ci_type = ci_type,
                             ci_level = ci_level, parallel = parallel, cores = cores, na.rm = na.rm)
      description <- data.table::data.table(description[1], description[2], description[3])
      names(description) <- c(st, "lower", "upper")
    }
  } else if(!is.data.frame(data)) {
    stop("data must either be a numeric vector or data frame")
  } else if(is.data.frame(data)){
    if(!is.numeric(dt[[y_str]])){
      stop("y must be a numeric vector or column of a data frame")
    }
    if(!missing(...)){
      g <- group_parser(data, ...)

      if(st == "mean") {
        description <- dt[,
                          .(measure = factor(c(st, "lower", "upper"), levels = c(st, "lower", "upper")),
                            value = mean_ci(get(y_str), replicates = replicates, ci_type = ci_type,
                                            ci_level = ci_level, parallel = parallel, cores = cores, na.rm = na.rm)),
                          by = eval(g)]
        description <- data.table::dcast(stats::na.omit(description), formula = ... ~ measure, value.var = "value")
      } else {
        description <- dt[,
                          .(measure = factor(c(st, "lower", "upper"), levels = c(st, "lower", "upper")),
                            value = stat_ci(get(y_str), stat = stat, replicates = replicates, ci_type = ci_type,
                                            ci_level = ci_level, parallel = parallel, cores = cores, na.rm = na.rm)),
                          by = eval(g)]
        description <- data.table::dcast(stats::na.omit(description), formula = ... ~ measure, value.var = "value")
      }
    } else {
      if(st == "mean") {
        description <- mean_ci(dt[[y_str]], replicates = replicates, ci_type = ci_type,
                               ci_level = ci_level, parallel = parallel, cores = cores, na.rm = na.rm)
        description <- data.table::data.table(description[1], description[2], description[3])
        names(description) <- c(st, "lower", "upper")
      } else {
        description <- stat_ci(dt[[y_str]], stat = stat, replicates = replicates, ci_type = ci_type,
                  ci_level = ci_level, parallel = parallel, cores = cores, na.rm = na.rm)
        description <- data.table::data.table(description[1], description[2], description[3])
        names(description) <- c(st, "lower", "upper")
      }
    }
  }
  if(output == "tibble") {
    description <- tibble::as_tibble(description)
  }
  return(description)
}

# describe_ci_all ---------------------------------------------------------
#' @title
#' Confidence intervals for a summary statistic of numeric variables in a data
#' frame.
#'
#' @description describe_ci_all extends the functionality of
#'   \code{\link{describe_ci}} by using the power of \code{\link[purrr]{map}} to
#'   enable you to obtain confidence intervals for a chosen summary statistic
#'   for each numeric variable in a data frame split by any number of grouping
#'   variables. Like \code{\link{stat_ci}} you can specify any function that
#'   operates on a numeric variable and returns a single value (e.g. mean,
#'   median, sd, se, etc.). Calculations of confidence intervals for the mean
#'   are obtained based on reference to the theoretical normal/gaussian
#'   distribution for speed, otherwise bootstrapping is used, with options for
#'   multicore machines to use parallel processing which can speed things up
#'   quite a bit for larger samples. \code{\link{stat_ci}} may be useful instead
#'   of describe_ci if you need to pass additional arguments to the chosen
#'   summary statistic function (which is what that function uses the ...
#'   argument for). To get confidence intervals for a single numeric variable in
#'   a data frame, use \code{\link{describe_ci}} instead.
#'
#' @importFrom dplyr select
#' @importFrom dplyr select_if
#' @importFrom dplyr arrange
#' @importFrom dplyr mutate
#' @importFrom dplyr group_by
#' @importFrom dplyr across
#' @importFrom dplyr all_of
#' @importFrom tidyr nest
#' @importFrom tidyr unnest
#' @importFrom purrr map
#' @importFrom tidyselect everything
#' @importFrom data.table data.table
#' @importFrom data.table as.data.table
#' @importFrom data.table rbindlist
#' @importFrom tibble as_tibble
#'
#' @param data A data frame or tibble containing the numeric vectors to be
#'   described and any grouping variables ("...").
#'
#' @param ... If the data object is a data.frame, this special argument accepts
#'   any number of unquoted grouping variable names (also present in the data source)
#'   to use for subsetting, separated by commas (e.g. \code{group_var1,
#'   group_var2})
#'
#' @param stat the unquoted name (e.g. mean, not "mean") of a summary statistic
#'   function to calculate confidence intervals for. Only functions which return
#'   a single value and operate on numeric variables are currently supported.
#'
#' @param replicates The number of bootstrap replicates to use to construct
#'   confidence intervals for statistics other than the sample mean. Default is
#'   2,000, as recommended by Efron & Tibshirani (1993). For publications, or if
#'   you need more precise estimates, more replications (e.g. >= 5,000) are
#'   recommended. N.B. more replications will of course take longer to run. If
#'   you get the error: "estimated adjustment 'a' is NA" when ci_type is set to
#'   "bca" then try again with more replications.
#'
#' @param ci_level The confidence level to use for constructing confidence
#'   intervals. Default is set to \code{ci_level = 0.95} for 95 percent CIs.
#'
#' @param ci_type The type of confidence intervals to calculate from the
#'   bootstrap samples. Most of the options available in the underlying boot.ci
#'   function are implemented (except for studentized intervals): "norm" for an
#'   approximation based on the normal distribution, "perc" for percentile,
#'   "basic" for basic, and "bca" for bias-corrected and accelerated. Percentile
#'   intervals are the default since these are typically sufficient when working
#'   with large data sets (e.g. >= 100,000 rows of data) and are faster to
#'   calculate than BCa intervals. However, BCa intervals (the default for the
#'   more primitive \code{\link{stat_ci}} function) tend to provide the most
#'   accurate/least-biased results (Efron, 1987), particularly for small-medium
#'   sized samples, at the obvious cost of requiring more time to calculate. See
#'   \code{\link[boot]{boot.ci}} for details.
#'
#' @param parallel set to TRUE if you want to use multiple cores or FALSE if you
#'   don't (the default). Note that there is some processing overhead involved
#'   when operating in parallel so speed gains may not be very noticeable for
#'   smaller samples (and may even take longer than sequential processing). Due
#'   to the nature of the underlying parallelization architecture, performance
#'   gains will likely be greater on non-Windows machines that can use the
#'   "multicore" implementation instead of "snow". For obvious reasons this
#'   option only works on machines with more than 1 logical processing core.
#'
#' @param cores If parallel is set to TRUE, this determines the number of cores
#'   to use. To see how many cores are available on your machine, use
#'   parallel::detectCores(). If cores is unspecified the number of available
#'   cores - 1 will be used by default.
#'
#' @param na.rm should missing values be removed before attempting to calculate
#'   the chosen statistic and confidence intervals? Default is TRUE.
#'
#' @param output "tibble" for tibble or "dt" for data.table. Tibble is used as
#'   the default output to facilitate subsequent use/modification of the output
#'   with the tidyverse collection of packages.
#'
#' @author Craig P. Hutton, \email{craig.hutton@@gov.bc.ca}
#'
#' @examples
#'
#' #using a single core (sequential processing)
#' \dontrun{
#' describe_ci_all(pdata[1:1000, ], stat = median) #bootstrapped CIs for the median
#' }
#' describe_ci_all(pdata, stat = mean) #the default
#' describe_ci_all(pdata, high_low, stat = mean) #split by a grouping variable
#'
#' \dontrun{
#' #using multiple cores (parallel processing)
#' describe_ci_all(pdata[1:1000, ], stat = sd, parallel = TRUE, cores = 2)
#' }
#'
#' @references
#' Efron, B. (1987). Better bootstrap confidence intervals. Journal of the
#' American statistical Association, 82(397), 171-185.
#'
#' Efron, B., & Tibshirani, R. J. (1993). An introduction to the bootstrap. New
#' York: Chapman & Hall.
#'
#' @seealso \code{\link[boot]{boot}}, \code{\link[boot]{boot.ci}},
#'   \code{\link{mean_ci}}, \code{\link{median_ci}}, \code{\link{stat_ci}},
#'   \code{\link{describe_ci}}
#'
#' @export
describe_ci_all <- function(data, ..., stat = mean, replicates = 2000,
                            ci_level = 0.95, ci_type = c("perc", "bca", "basic", "norm"),
                            parallel = FALSE, cores = NULL,
                            na.rm = TRUE, output = c("tibble", "dt")) {

  ci_type <- match.arg(ci_type)
  output <- match.arg(output)

  st <- deparse(substitute(stat))
  dscr_ci_all <- function(data, stat = mean, replicates = 2000,
                          ci_level = 0.95, ci_type = "perc", parallel = FALSE, cores = NULL,
                          na.rm = TRUE, output = "dt") {
    num_data <- dplyr::select_if(data, is.numeric)
    st_ci <- function(data, stat = mean, replicates = 2000,
                      ci_level = 0.95, ci_type = "perc", parallel = FALSE, cores = NULL,
                      na.rm = TRUE, output = "dt"){
      dt <- data.table::as.data.table(data)
      if(st == "mean") {
        description <- mean_ci(dt[[1]])
        description <- data.table::data.table(description[1], description[2], description[3])
        names(description) <- c(st, "lower", "upper")
      } else {
        description <- stat_ci(dt[[1]], stat = stat, replicates = replicates, ci_type = ci_type,
                  ci_level = ci_level, parallel = parallel, cores = cores, na.rm = na.rm)
        description <- data.table::data.table(description[1], description[2], description[3])
        names(description) <- c(st, "lower", "upper")
      }
      return(description)
    }
    if(st == "mean"){
      out <- purrr::map(num_data,
                   ~st_ci(data = .x, stat = mean, replicates = replicates, ci_type = ci_type,
                          ci_level = ci_level, parallel = parallel, cores = cores, na.rm = na.rm,
                          output = "dt"))
      out <- data.table::rbindlist(out, use.names = TRUE, idcol = "variable", fill = TRUE)
    } else {
      out <- purrr::map(num_data,
                   ~st_ci(data = .x, stat = stat, replicates = replicates, ci_type = ci_type,
                          ci_level = ci_level, parallel = parallel, cores = cores, na.rm = na.rm,
                          output = "dt"))
      out <- data.table::rbindlist(out, use.names = TRUE, idcol = "variable", fill = TRUE)
    }
    return(out)
  }

  if(!missing(...)){
    g <- group_parser(data, ...)
    description <- dplyr::group_by(data, dplyr::across(dplyr::all_of(g)))
    description <- tidyr::nest(description)
    description <- dplyr::mutate(description,
                                 data = purrr::map(data,
                                                   ~dscr_ci_all(data = .x, stat = stat, replicates = replicates,
                                                                ci_level = ci_level, ci_type = ci_type,
                                                                parallel = parallel, cores = cores, na.rm = na.rm)))
    description <- tidyr::unnest(description, data)
    description <- dplyr::select(description, variable, tidyselect::everything())
    description <- dplyr::arrange(description, variable)
  } else {
    description <- dscr_ci_all(data, stat = stat, replicates = replicates,
                               ci_level = ci_level, ci_type = ci_type,
                               parallel = parallel, cores = cores, na.rm = na.rm)
  }
  if(output == "tibble") {
    description <- tibble::as_tibble(description)
  } else if(output == "dt"){
    description <- data.table::as.data.table(description)
  }
  return(description)
}
